### PR TITLE
(4.x) Merge 3.4

### PR DESCRIPTION
**Pull request by alalek**
#3174 from AleksandrPanov:fix_gridboard_objPoints
#3181 from sturkmen72:bug-fix-edge_drawing
#3186 from AleksandrPanov:fix_CORNER_REFINE_CONTOUR
#3201 from AleksandrPanov:fix_filterTooCloseCandidates

Main PR: https://github.com/opencv/opencv/pull/21784
Previous "Merge 3.4": #3173

<cut/>

<details>

```
force_builders=Custom,Win64 OpenCL
buildworker:Custom=linux-4,linux-6
build_image:Custom=ubuntu-cuda:18.04
buildworker:Win64 OpenCL=windows-2
build_image:Win64 OpenCL=msvs2019
```

</details>

### CODE REVIEW

File: modules/aruco/include/opencv2/aruco/charuco.hpp [hunk 1]
```diff
@@ -285,7 +287,8 @@ CV_EXPORTS_W void detectCharucoDiamond(InputArray image, InputArrayOfArrays mark
                                        InputArray markerIds, float squareMarkerLengthRate,
                                        OutputArrayOfArrays diamondCorners, OutputArray diamondIds,
                                        InputArray cameraMatrix = noArray(),
-                                       InputArray distCoeffs = noArray());
+                                       InputArray distCoeffs = noArray(),
+                                       Ptr<Dictionary> dictionary = getPredefinedDictionary(PREDEFINED_DICTIONARY_NAME(0)));
```
Comment (alalek) [inline modules/aruco/include/opencv2/aruco/charuco.hpp:291]: > `Ptr<Dictionary>`

@AleksandrPanov Why is not a const reference?

File: modules/aruco/test/test_charucodetection.cpp [hunk 2]
```diff
@@ -678,7 +611,7 @@ TEST(Charuco, testCharucoCornersCollinear_false)
 }
 
 // test that ChArUco board detection is subpixel accurate
-TEST(Charuco, testBoardSubpixelCoords)
+TEST(Charuco, DISABLED_testBoardSubpixelCoords)  // FIXIT: https://github.com/opencv/opencv_contrib/pull/3213
```
Comment (alalek) [inline modules/aruco/test/test_charucodetection.cpp:614]: @AleksandrPanov Disabled this test. It is completely rewritten in 4.x branch (see #3213). Main question is "why?"

**Comment by alalek** [comment]
:+1:

File: modules/aruco/include/opencv2/aruco/charuco.hpp (hunk 1 above)
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco/charuco.hpp]: `const Dictionary&` does not work for python binding.
Or do you suggest `const Ptr<Dictionary>&`?

**Pull request merged**